### PR TITLE
feat: add `allow` option for `restrict-template-expressions`

### LINKED ISSUE #6279: Enhancement: [restrict-template-expressions] More permissive type check
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/restrict-template-expressions/

### Description

Related: #6203 #5325

I've originally opened #5600 because I'd like to leverage this rule to discover the *accidental misuse* of string template expression (e.g., the mostly occurring case is, interpolating a `() => string`, `Promise<string>` or `Promise<...>` instead of `string`). However, after I've opened `restrict-template-expressions`. there seems to be a lot of rule violations that cannot be silenced with existing options. For example,
* `string[]`
* `unknown`
* `never`
* `URL`
* `URLSearchParameters`

Instead of defining what types are *allowed* in the interpolate expression, is it possible to set up a list of blocked types?

### Fail

```typescript
const formatErrorMessage = (error: unknown) => String(error);
// `formatErrorMessage` is of function type. Converting it into string results in function body being printed.
const message = "Error is {formatErrorMessage}";
```

### Pass

```typescript
const formatErrorMessage = (error: unknown) => String(Error);
// Call the function to return the string result.
const message1 = "Error is {formatErrorMessage(err)}.";
// Or if you are converting function into string by design.
const message2 = "Error is {String(formatErrorMessage)}.";

// Stringify the URL
let url: URL | undefined;
const message3 = "URL is {url}.";
```

### Additional Info

_No response_

### LINKED ISSUE #8722: Enhancement: [restrict-template-expressions] More permissive type check
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/restrict-template-expressions/

### Description

Related: #6279

I have added the `allowArray` option to `restrict-template-expressions` in #8389.

Now, I'd like to discuss other things that may be useful to put in a string. The most common case for me is stringifying `URLSearchParams`, `Date` and `Buffer`s, but there could be potential for others as well.
Some questions for discussion:

- Do we add a new option `ignoredTypeNames`? (I have created a draft PR #8556 implementing this)
- What will be the default value for this option?
- What does this mean for the `no-base-to-string` rule, which is becoming increasingly similar to this rule?

### Fail

```TypeScript
declare const maybeParams: URLSearchParams | undefined;

fetch(`https://myapi.com?${maybeParams}`);
```

### Pass

```TypeScript
declare const definitelyParams: URLSearchParams;

fetch(`https://myapi.com?${definitelyParams}`);
```

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by abrahamguo**
## PR Checklist

- [x] Addresses an existing open issue: fixes #8722
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview
Building on the work from [#8389](https://github.com/typescript-eslint/typescript-eslint/pull/8389), which @JoshuaKGoldberg has been reviewing, this PR ports the `ignoredTypeNames` option from `no-base-to-string` into `restrict-template-expressions`. I'm open to further discussion about this, as this update definitely blurs the line between the two rules.

### CODE REVIEW

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Thanks for sending this in - but would you be open to filing an issue? I think we'd want to discuss this a bit before landing it. 

Also, just a couple preliminary review thoughts if it does land:
* Unit tests?
* I think we'd want to share the default value of `ignoredTypeNames` in a constant so it doesn't risk getting out of sync

**Comment by JoshuaKGoldberg** [comment]
Marking as a draft pending #8722's resolution.

**Comment by JoshuaKGoldberg** [comment]
#8722 marked as accepting PRs, so un-drafting this. Thanks for your patience with us @abrahamguo!

File: packages/eslint-plugin/src/rules/restrict-template-expressions.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/restrict-template-expressions.ts:54]: 🤔 the other options are named like `allow*`. Other places that use the standard format:

* https://typescript-eslint.io/rules/prefer-readonly-parameter-types#allow: `allow`
* https://github.com/typescript-eslint/typescript-eslint/pull/9186: `allowForKnownSafePromises`

Neither of those options mention "type" or "name", because it's redundant in the options' contexts. I'd vote for this rule's option being just `allow`.

Comment (abrahamguo) [reply packages/eslint-plugin/src/rules/restrict-template-expressions.ts:54]: I'm fine with this. However, I will point out that the option in [`no-base-to-string`](https://typescript-eslint.io/rules/no-base-to-string) is called `ignoredTypeNames`, and `no-base-to-string` is a much more similar rule to this one than the two rules you've mentioned above.

**Comment by abrahamguo** [comment]
Work done:
- [x] Since `TypeOrValueSpecifier` is now available in 3 rules, extract the documentation to a shared file
- `no-floating-promises.allowForKnownSafePromises`
- `prefer-readonly-parameter-types.allow`
- `restrict-template-expressions.allow`
- [x] Add documentation for the array form of `TypeOrValueSpecifier.name`, which was previously undocumented.
- [x] Extract a little bit more shared code into `TypeOrValueSpecifier.ts` to make it easier for more rules to implement this
- [x] Discover and fix bug where `URL` and `URLSearchParams` were not able to be specified with `TypeOrValueSpecifier`.
- In this repo (and many repos), these types come from `@types/node`.
- `program.isSourceFileDefaultLibrary()` returns `false`, meaning that `{ from: lib }` does not work
- `program.sourceFileToPackageName.get()` returns `undefined`, meaning that `{ from: package }` does not work
- Even though technically these types come from a package, (A) they are not explicitly imported, and (B) many users may have `@types/node` as a transitive dependency, so I think it makes sense to special-case types from `@types/{bun|node}`.
- [x] Implement `restrict-template-expressions.allow`.
- Default value is `[Error, URL, URLSearchParams]`, matching `no-base-to-string`. All of these produce a useful result when stringified.

Items for consideration:

- [ ] This rule is getting more similar to `no-base-to-string`. Is that OK? Do we combine the rules?
- [ ] Naming this option:
- Should it match the option naming pattern of the other two rules that use `TypeOrValueSpecifier` (`allow`)?
- Should it match the option name in `no-base-to-string` (`ignoredTypeNames`), with which it shares a default value, and since it is so similar to that rule? (But, on the other hand, `no-base-to-string.ignoredTypeNames` is `string[]`, whereas this is `TypeOrValueSpecifier[]`)
- [ ] In this PR, should we also move `no-base-to-string` to `TypeOrValueSpecifier[]`? This would not be a breaking change, since `string[]` is also `TypeOrValueSpecifier[]`.

**Comment by JoshuaKGoldberg** [comment]
Great progress! I'm marking this as a draft & unsubscribing to help my beleaguered inbox - but feel free to ping if you any blockers come up, and/or un-draft once it's ready. Cheers!

**Comment by abrahamguo** [comment]
@JoshuaKGoldberg Sorry to not be clear! This is ready for review — I'd like your input on the remaining items above 👍

File: packages/eslint-plugin/docs/rules/shared/TypeOrValueSpecifier.md
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/docs/rules/shared/TypeOrValueSpecifier.md:1]: [Praise] 👍 I like this! This'll be a good stepping stone to a more formal docs page.

File: packages/eslint-plugin/tests/docs.test.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/docs.test.ts:1]: [Style] There are a lot of changes to this file that are just newlines. It's a little harder to review PRs when a lot of lines are falsely being reported as changed. Could you revert them please?

**Comment by JoshuaKGoldberg** [comment]
Oh gotcha! I made individual threads for each of the open question, in case we want to discuss them more deeply.

File: packages/eslint-plugin/docs/rules/restrict-template-expressions.mdx
Comment (abrahamguo) [reply packages/eslint-plugin/docs/rules/restrict-template-expressions.mdx:130]: Makes sense!

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Getting there! I'm happy to see how straightforward the code changes to the actual rule are (ignoring the `.shot` additions, heh). ⚡

File: packages/eslint-plugin/tests/rules/restrict-template-expressions.test.ts
Comment (abrahamguo) [reply packages/eslint-plugin/tests/rules/restrict-template-expressions.test.ts:320]: I'm guessing you meant `from: 'lib'`, since `Promise` comes from `lib` rather than `file`

File: packages/type-utils/src/TypeOrValueSpecifier.ts
Comment (abrahamguo) [reply packages/type-utils/src/TypeOrValueSpecifier.ts:197]: Waiting on more info from https://github.com/typescript-eslint/typescript-eslint/pull/9500/files#r1669456255

Comment (JoshuaKGoldberg) [reply packages/type-utils/src/TypeOrValueSpecifier.ts:197]: Okie now that #9500 is merged, I think this is unblocked @abrahamguo? Are you waiting on us for anything else?

Comment (abrahamguo) [reply packages/type-utils/src/TypeOrValueSpecifier.ts:197]: Yes, looks like #9500 did not improve the issues I was targeting here.
I've opened #9608 for `URL` and I'll open another issue in the next day or so for `URLSearchParams`.

Comment (abrahamguo) [reply packages/type-utils/src/TypeOrValueSpecifier.ts]: After some investigation, it looks like `URL` and `URLSearchParams` have the same issue. It _is_ an issue, but only when `DOM` is _not_ in the `lib`. This means that in the real world most of the time, it will not be a problem — just in our integration test, which I'll fix. So #9608 shouldn't be a blocker, and I'll get this PR wrapped up!

File: packages/eslint-plugin/docs/rules/restrict-template-expressions.mdx
Comment (abrahamguo) [reply packages/eslint-plugin/docs/rules/restrict-template-expressions.mdx:1]: Moved to #9644

**Comment by abrahamguo** [comment]
Note that #9648 blocks any references to `URL` or `URLSearchParams` in unit tests or documentation code examples, so I either have to:

1. Have #9648 accepted and fixed so that I am able to use references to `URL` and `URLSearchParams` in docs examples and unit tests, or
2. omit `URL` and `URLSearchParams` from docs examples and unit tests, which means that even though the default value is `{ allow: [{ from: 'lib', name: 'Error' }, { from: 'lib', name: 'URL' }, { from: 'lib', name: 'URLSearchParams' }] }`, I can only test, and show examples for, `Error`, since it is the only one of the three that does not depend on `lib: DOM`.

**Comment by JoshuaKGoldberg** [comment]
> 9648 blocks any references to `URL` or `URLSearchParams`

`RuleTester` can manually specify `languageOptions.parserOptions.lib`, no?

**Review by JoshuaKGoldberg — DISMISSED** [review]
This is fantastic, great work @abrahamguo! Really clean implementation and I like the sharing of docs. Thanks!

I'll also want to wait on another review from @typescript-eslint/triage-team. The more we use the shared _TypeOrValueSpecifier_ format the more review eyes we'll want on it.

IMO we should land this before #9875 and then I can adjust the docs there to fit this PR.

Captain Holt in Brooklyn 99 Happily clapping his hands together and shouting 'Hot damn!' with fiery steel caption font

**Review by kirkwaiblinger — COMMENTED** [review]
Since https://github.com/typescript-eslint/typescript-eslint/pull/9875/files ended up landing first, some merge conflicts will need to be addressed. Ping for another review pass after that is completed?

**Comment by abrahamguo** [comment]
Conflicts resolved and ready for re-review!

I am not clear why the one test is failing. It works locally for me.

**Comment by JoshuaKGoldberg** [comment]
It's from the CI job that runs tests with the project service enabled:

https://github.com/typescript-eslint/typescript-eslint/blob/45b7d810f61ce752c1dcb4b6bb4f3cc8357492cf/.github/workflows/ci.yml#L247

Copying the failure here:

```plaintext
  ● restrict-template-expressions › valid › const msg = `arg = ${new URLSearchParams()}`;

    assert.strictEqual(received, expected)

    Expected value to strictly be equal to:
      0
    Received:
      1
    
    Message:
      Should have no errors but had 1: [
      {
        ruleId: '@rule-tester/restrict-template-expressions',
        severity: 2,
        message: 'Invalid type "URLSearchParams" of template literal expression.',
        line: 1,
        column: 22,
        nodeType: 'NewExpression',
        messageId: 'invalidType',
        endLine: 1,
        endColumn: 43
      }
    ]
```

You can directly test that file in `packages/eslint-plugin` with `TYPESCRIPT_ESLINT_PROJECT_SERVICE=true yarn jest restrict-template-`.

**Comment by abrahamguo** [comment]
Hmm, looks like setting `TYPESCRIPT_ESLINT_PROJECT_SERVICE=true` causes it to bypass the `tsconfig-withdom.json` set in `restrict-template-expressions.test.ts`.

**Comment by abrahamguo** [comment]
I've worked around this by ensuring that (similar to the new examples in the docs page) the unit tests only refer to the `Error` type, which is available with or without the `DOM` types.
`URL` and `URLSearchParams` are not shown in the docs code examples, nor in the unit tests, since they require `DOM` types in order to work properly.

**Review by JoshuaKGoldberg — APPROVED** [review]
✨

File: packages/eslint-plugin/docs/rules/restrict-template-expressions.mdx
Comment (kirkwaiblinger) [inline packages/eslint-plugin/docs/rules/restrict-template-expressions.mdx:88]: lol

**Review by kirkwaiblinger — APPROVED** [review]
Peachy keen!

**Pull request merged**
